### ISSUE TITLE
Create vue 3 component

### ISSUE THREAD

**Issue by mika76** [apexcharts/vue-apexcharts#221]
I converted this to a vue 3 component for my use and when you decide to create one I thought I'd give you a head start. Seems to work fine for me but it has not gone through any rigorous testing...

Also note I commented out the event listeners propagating - I couldn't figure it out but was not using it myself...

```js
const apexchart = defineComponent({
            props: {
                options: {
                    type: Object
                },
                type: {
                    type: String
                },
                series: {
                    type: Array,
                    required: true,
                    default: () => []
                },
                width: {
                    default: '100%'
                },
                height: {
                    default: 'auto'
                }
            },
            setup(props, ctx) {
                const el = ref(null);
                const chart = ref(null);

                const isObject = (item) => {
                    return (
                        item && typeof item === 'object' && !Array.isArray(item) && item != null
                    )
                };

                const extend = (target, source) => {
                    if (typeof Object.assign !== 'function') {
                        (function () {
                            Object.assign = function (target) {
                                // We must check against these specific cases.
                                if (target === undefined || target === null) {
                                    throw new TypeError('Cannot convert undefined or null to object')
                                }

                                let output = Object(target)
                                for (let index = 1; index < arguments.length; index++) {
                                    let source = arguments[index]
                                    if (source !== undefined && source !== null) {
                                        for (let nextKey in source) {
                                            if (source.hasOwnProperty(nextKey)) {
                                                output[nextKey] = source[nextKey]
                                            }
                                        }
                                    }
                                }
                                return output
                            }
                        })()
                    }

                    let output = Object.assign({}, target)
                    if (isObject(target) && isObject(source)) {
                        Object.keys(source).forEach((key) => {
                            if (isObject(source[key])) {
                                if (!(key in target)) {
                                    Object.assign(output, {
                                        [key]: source[key]
                                    })
                                } else {
                                    output[key] = extend(target[key], source[key])
                                }
                            } else {
                                Object.assign(output, {
                                    [key]: source[key]
                                })
                            }
                        })
                    }
                    return output
                };

                const init = () => {
                    const newOptions = {
                        chart: {
                            type: props.type || props.options.chart.type || 'line',
                            height: props.height,
                            width: props.width,
                            events: {}
                        },
                        series: props.series
                    }

                    // Object.keys(ctx.listeners).forEach((evt) => {
                    //     newOptions.chart.value.events[evt] = ctx.listeners[evt];
                    // });

                    const config = extend(props.options, newOptions);
                    chart.value = new ApexCharts(el.value, config)
                    return chart.value.render()
                };

                const refresh = () => {
                    destroy();
                    return init();
                };

                const destroy = () => {
                    chart.value.destroy()
                };

                const updateSeries = (newSeries, animate) => {
                    return chart.value.updateSeries(newSeries, animate)
                };

                const updateOptions = (newOptions, redrawPaths, animate, updateSyncedCharts) => {
                    return chart.value.updateOptions(newOptions, redrawPaths, animate, updateSyncedCharts)
                };

                const toggleSeries = (seriesName) => {
                    return chart.value.toggleSeries(seriesName)
                };

                const showSeries = (seriesName) => {
                    chart.value.showSeries(seriesName)
                };

                const hideSeries = (seriesName) => {
                    chart.value.hideSeries(seriesName)
                };

                const appendSeries = (newSeries, animate) => {
                    return chart.value.appendSeries(newSeries, animate)
                };

                const resetSeries = () => {
                    chart.value.resetSeries()
                };

                const toggleDataPointSelection = (seriesIndex, dataPointIndex) => {
                    chart.value.toggleDataPointSelection(seriesIndex, dataPointIndex)
                };

                const appendData = (newData) => {
                    return chart.value.appendData(newData)
                };

                const addText = (options) => {
                    chart.value.addText(options)
                };

                const dataURI = () => {
                    return chart.value.dataURI()
                };

                const setLocale = (localeName) => {
                    return chart.value.setLocale(localeName)
                };

                const addXaxisAnnotation = (options, pushToMemory) => {
                    chart.value.addXaxisAnnotation(options, pushToMemory)
                };

                const addYaxisAnnotation = (options, pushToMemory) => {
                    chart.value.addYaxisAnnotation(options, pushToMemory)
                };

                const addPointAnnotation = (options, pushToMemory) => {
                    chart.value.addPointAnnotation(options, pushToMemory)
                };

                const removeAnnotation = (id, options) => {
                    chart.value.removeAnnotation(id, options)
                };

                const clearAnnotations = () => {
                    chart.value.clearAnnotations()
                };

                onMounted(() => {
                    init();
                });

                onBeforeUnmount(() => {
                    if (!chart.value) {
                        return
                    }
                    destroy()
                });

                watch(() => props.options, () => {
                    if (!chart.value && props.options) {
                        init()
                    } else {
                        chart.value.updateOptions(props.options)
                    }
                });

                watch(() => props.series, () => {
                    if (!chart.value && props.series) {
                        init()
                    } else {
                        chart.value.updateSeries(props.series)
                    }
                });

                watch(() => props.type, () => {
                    refresh();
                });

                watch(() => props.width, () => {
                    refresh();
                });

                watch(() => props.height, () => {
                    refresh();
                });

                return () =>
                    h("div", {
                        ref: el
                    });
            },
        });
```

**Comment by junedchhipa** [comment]
Awesome!
Haven't looked into Vue 3 yet, but do you think the component can be made to support both versions (2 and 3)?
Thanks for contributing.

**Comment by mika76** [comment]
Honestly I don't know - I've just started looking into it myself. I'll let you know when I figure it out 😄

**Comment by rsornellas** [comment]
example link?

I am trying to use ApexChart with Vue3 and I have this error

![image](https://user-images.githubusercontent.com/33547753/90837238-dbdaca00-e327-11ea-879c-d55ecf5bc675.png)

**Comment by mika76** [comment]
Hey @rsornellas I'm using the above still in a private project and it seems to be working for me - maybe it's the version of apex charts? Or your calling component?

**Comment by maharzan** [comment]
@mika76 Any simple example like barchart would be really helpful. Thanks.

**Comment by mika76** [comment]
So here's my full component, I notice the one difference is the `nextTick` which waits for the element to be placed in the dom. To use, just import the component and use as in the vue 2 examples. I'm currently using wit vue rc 10 and it seems to be ok.

```vue
<template>
  <div ref="el"></div>
</template>
<script>
import {
  defineComponent,
  ref,
  onMounted,
  onBeforeUnmount,
  watch,
  h,
  onBeforeMount,
  nextTick,
} from "vue";
import ApexCharts from "apexcharts";

export default defineComponent({
  props: {
    options: {
      type: Object,
    },
    type: {
      type: String,
    },
    series: {
      type: Array,
      required: true,
      default: () => [],
    },
    width: {
      default: "100%",
    },
    height: {
      default: "auto",
    },
  },
  setup(props, ctx) {
    const el = ref(null);
    const chart = ref(null);

    const isObject = (item) => {
      return (
        item && typeof item === "object" && !Array.isArray(item) && item != null
      );
    };

    const extend = (target, source) => {
      if (typeof Object.assign !== "function") {
        (function () {
          Object.assign = function (target) {
            // We must check against these specific cases.
            if (target === undefined || target === null) {
              throw new TypeError("Cannot convert undefined or null to object");
            }

            let output = Object(target);
            for (let index = 1; index < arguments.length; index++) {
              let source = arguments[index];
              if (source !== undefined && source !== null) {
                for (let nextKey in source) {
                  if (source.hasOwnProperty(nextKey)) {
                    output[nextKey] = source[nextKey];
                  }
                }
              }
            }
            return output;
          };
        })();
      }

      let output = Object.assign({}, target);
      if (isObject(target) && isObject(source)) {
        Object.keys(source).forEach((key) => {
          if (isObject(source[key])) {
            if (!(key in target)) {
              Object.assign(output, {
                [key]: source[key],
              });
            } else {
              output[key] = extend(target[key], source[key]);
            }
          } else {
            Object.assign(output, {
              [key]: source[key],
            });
          }
        });
      }
      return output;
    };

    const init = async () => {
      await nextTick();

      const newOptions = {
        chart: {
          type: props.type || props.options.chart.type || "line",
          height: props.height,
          width: props.width,
          events: {},
        },
        series: props.series,
      };

      // Object.keys(ctx.listeners).forEach((evt) => {
      //     newOptions.chart.value.events[evt] = ctx.listeners[evt];
      // });

      const config = extend(props.options, newOptions);
      chart.value = new ApexCharts(el.value, config);
      return chart.value.render();
    };

    const refresh = () => {
      destroy();
      return init();
    };

    const destroy = () => {
      chart.value.destroy();
    };

    const updateSeries = (newSeries, animate) => {
      return chart.value.updateSeries(newSeries, animate);
    };

    const updateOptions = (
      newOptions,
      redrawPaths,
      animate,
      updateSyncedCharts
    ) => {
      return chart.value.updateOptions(
        newOptions,
        redrawPaths,
        animate,
        updateSyncedCharts
      );
    };

    const toggleSeries = (seriesName) => {
      return chart.value.toggleSeries(seriesName);
    };

    const showSeries = (seriesName) => {
      chart.value.showSeries(seriesName);
    };

    const hideSeries = (seriesName) => {
      chart.value.hideSeries(seriesName);
    };

    const appendSeries = (newSeries, animate) => {
      return chart.value.appendSeries(newSeries, animate);
    };

    const resetSeries = () => {
      chart.value.resetSeries();
    };

    const toggleDataPointSelection = (seriesIndex, dataPointIndex) => {
      chart.value.toggleDataPointSelection(seriesIndex, dataPointIndex);
    };

    const appendData = (newData) => {
      return chart.value.appendData(newData);
    };

    const addText = (options) => {
      chart.value.addText(options);
    };

    const dataURI = () => {
      return chart.value.dataURI();
    };

    const setLocale = (localeName) => {
      return chart.value.setLocale(localeName);
    };

    const addXaxisAnnotation = (options, pushToMemory) => {
      chart.value.addXaxisAnnotation(options, pushToMemory);
    };

    const addYaxisAnnotation = (options, pushToMemory) => {
      chart.value.addYaxisAnnotation(options, pushToMemory);
    };

    const addPointAnnotation = (options, pushToMemory) => {
      chart.value.addPointAnnotation(options, pushToMemory);
    };

    const removeAnnotation = (id, options) => {
      chart.value.removeAnnotation(id, options);
    };

    const clearAnnotations = () => {
      chart.value.clearAnnotations();
    };

    onBeforeMount(() => {
      window.ApexCharts = ApexCharts;
    });

    onMounted(() => {
      init();
    });

    onBeforeUnmount(() => {
      if (!chart.value) {
        return;
      }
      destroy();
    });

    watch(
      () => props.options,
      () => {
        if (!chart.value && props.options) {
          init();
        } else {
          chart.value.updateOptions(props.options);
        }
      }
    );

    watch(
      () => props.series,
      () => {
        if (!chart.value && props.series) {
          init();
        } else {
          chart.value.updateSeries(props.series);
        }
      }
    );

    watch(
      () => props.type,
      () => {
        refresh();
      }
    );

    watch(
      () => props.width,
      () => {
        refresh();
      }
    );

    watch(
      () => props.height,
      () => {
        refresh();
      }
    );

    return { el };
  },
});
</script>
```

**Comment by jd-0001** [comment]
Awaiting for Vue 3 Support :eyes:

**Comment by junedchhipa** [comment]
@mika76 Thanks for the work on this.
I will try this out soon and I plan to release `vue3-apexcharts` once everything works out fine.

**Comment by junedchhipa** [comment]
@mika76 I have created a new repository - [vue3-apexcharts](https://github.com/apexcharts/vue3-apexcharts). I would appreciate it if you can take a look and see if everything works.

**Comment by mika76** [comment]
@junedchhipa sure I'll try it as soon as I can

**Comment by junedchhipa** [comment]
Released a new version [vue3-apexcharts@1.1.0](https://www.npmjs.com/package/vue3-apexcharts) which is compatible with Vue 3.0
The [repository](https://github.com/apexcharts/vue3-apexcharts) contains a demo also to show how to use it.
